### DISCUSSION TITLE
text wrapping for application output

### DISCUSSION THREAD

**Discussion by breed** [remkop/picocli#1753] [Show and tell]
the Ansi class allows for applications to colorize the output, but is there also a way to leverage the text wrapping that picocli has so that my output will wrap properly at screen boundaries?

**Comment by remkop** [comment]
I am not sure what "properly" means for you: if you just print a very very long string, it will wrap automatically at screen boundaries.
You probably don't mean that. But not sure what it is that you want instead...

**Reply by breed** [reply]
yes, i would like it to word wrap at the screen boundary. the usage text word wraps (the word is not split at the boundary). System.out.println(Ansi.AUTO.string("very long sentence ... words.")) does will continue the string on the next line but it make break a word between two lines.

**Reply by remkop** [reply]
Thank you for the clarification.
There are two aspects to this: how to get the screen width, and how to break text at word boundaries.

How to get the screen width: picocli's API has the [UsageMessageSpec.width](https://picocli.info/apidocs-all/info.picocli/picocli/CommandLine.Model.UsageMessageSpec.html#width()) method. This returns the terminal width if the spec's [autowidth](https://picocli.info/apidocs-all/info.picocli/picocli/CommandLine.Model.UsageMessageSpec.html#autoWidth()) is set to true (or when system property `picocli.usage.width` is set to `AUTO`).

To break text at word boundaries, you can either use picocli's [Help.TextTable](https://picocli.info/apidocs-all/info.picocli/picocli/CommandLine.Help.TextTable.html) class, or use a line break iterator `java.text.BreakIterator.getLineInstance()` to find the word boundaries and manually split up the string into parts that fit within the terminal width.

One idea is to do something like this:

```java
CommandSpec spec = myCommandLine.getCommandSpec();
spec.usageMessage().autoWidth(true); // use terminal width
int width = spec.usageMessage().width(); // obtain the terminal width

// indentation = how many leading spaces for the second and subsequent lines
// when text is wrapped over multiple lines (the first line is never indented)
int indent = 0; 
CommandLine.Help.Column line = new Column(width, indent, CommandLine.Help.Column.Overflow.WRAP);

// make a TextTable with just one column
CommandLine.Help.ColorScheme colorScheme = CommandLine.defaultColorScheme(CommandLine.Help.Ansi.AUTO);
CommandLine.Help.TextTable txtTable = CommandLine.Help.TextTable.forColumns(colorScheme, new Column[] {line});

// then call the `addRowValues` method 
txtTable.addRowValues("This is the first line that should be wrapped at word boundaries automatically. Feel free to make this line as long as you want and it should wrap to the next line with the specified indentation.");
txtTable.addRowValues("@|bold,green,underline Hello, colored world!|@ This is the second line that should be wrapped at word boundaries automatically...");
txtTable.addRowValues("This is the third line that should be wrapped at word boundaries automatically...");
```

**Reply by breed** [reply]
excellent! that is exactly what i was looking for. i found i also needed to add txtTable.indentWrappedLines = 0.

thank you for the excellent library!

ben
